homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,208 @@
+    """Representation of a Sonos Battery entity."""
+
+    def __init__(self, soco: SoCo, sonos_data: SonosData, battery_info: Dict[str, Any]):
+        """Initialize a SonosBatteryEntity."""
+        super().__init__(soco, sonos_data)
+        self._battery_info = battery_info
+
+    async def async_added_to_hass(self) -> None:
+        """Register polling callback when added to hass."""
+        self.async_on_remove(
+            self.hass.helpers.event.async_track_time_interval(
+                self.update, BATTERY_SCAN_INTERVAL
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, f"{SONOS_SEEN}-{self.soco.uid}", self.async_seen
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass, f"{SONOS_UNSEEN}-{self.soco.uid}", self.async_unseen
+            )
+        )
+        self.async_on_remove(
+            async_dispatcher_connect(
+                self.hass,
+                f"{SONOS_PROPERTIES_UPDATE}-{self.soco.uid}",
+                self.async_update_battery_info,
+            )
+        )
+
+    async def async_update_battery_info(self, event: SonosEvent = None) -> None:
+        """Update battery info using the provided SonosEvent."""
+        if event is None:
+            return
+
+        if (more_info := event.variables.get("more_info")) is None:
+            return
+
+        more_info_dict = dict(x.split(":") for x in more_info.split(","))
+
+        is_charging = EVENT_CHARGING[more_info_dict["BattChg"]]
+        if is_charging == self.charging:
+            self._battery_info.update({"Level": int(more_info_dict["BattPct"])})
+        else:
+            if battery_info := await self.hass.async_add_executor_job(
+                fetch_battery_info_or_none, self.soco
+            ):
+                self._battery_info = battery_info
+
+        self.schedule_update_ha_state()
+
+    async def async_seen(self, soco) -> None:
+        """Record that this player was seen right now."""
+        self.soco = soco
+        self.async_write_ha_state()
+
+    @callback
+    def async_unseen(self, now=None):
+        """Make this player unavailable when it was not seen recently."""
+        self.async_write_ha_state()
+
+    # Identification of this Entity
+    @property
+    def unique_id(self) -> str:
+        """Return the unique ID of the sensor."""
+        return f"{self.soco.uid}-battery"
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        speaker_info = self.data.speaker_info[self.soco.uid]
+        return f"{speaker_info['zone_name']} Battery"
+
+    @property
+    def device_class(self) -> str:
+        """Return the entity's device class."""
+        return DEVICE_CLASS_BATTERY
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Get the unit of measurement."""
+        return PERCENTAGE
+
+    # Update the current state
```

Comment:
Looks like `should_poll` being `False` is missing (probably good to move it into the base class)